data-prepper-plugins/opensearch-api-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearchapi/OpenSearchAPIService.java

```diff
@@ -0,0 +1,205 @@
+
+        //TODO: Will need to revisit the metrics per API endpoint
+        requestsReceivedCounter = pluginMetrics.counter(REQUESTS_RECEIVED);
+        successRequestsCounter = pluginMetrics.counter(SUCCESS_REQUESTS);
+        payloadSizeSummary = pluginMetrics.summary(PAYLOAD_SIZE);
+        requestProcessDuration = pluginMetrics.timer(REQUEST_PROCESS_DURATION);
+    }
+
+    @Post("/_bulk")
+    public HttpResponse doPostBulk(final ServiceRequestContext serviceRequestContext, final AggregatedHttpRequest aggregatedHttpRequest,
+                                   @Param("pipeline") @Nullable String pipeline,
+                                   @Param("routing") @Nullable String routing) throws Exception {
+
+        BulkAPIRequestParams bulkAPIRequestParams = BulkAPIRequestParams.builder()
+                .pipeline(pipeline)
+                .routing(routing)
+                .build();
+        return requestProcessDuration.recordCallable(() -> processBulkRequest(serviceRequestContext, aggregatedHttpRequest, bulkAPIRequestParams));
+    }
+
+    @Post("/{index}/_bulk")
+    public HttpResponse doPostBulkIndex(final ServiceRequestContext serviceRequestContext, final AggregatedHttpRequest aggregatedHttpRequest,
+                                        @Param("index") String index,
+                                        @Param("pipeline") @Nullable String pipeline,
+                                        @Param("routing") @Nullable String routing) throws Exception {
+        BulkAPIRequestParams bulkAPIRequestParams = BulkAPIRequestParams.builder()
+                .index(index)
+                .pipeline(pipeline)
+                .routing(routing)
+                .build();
+        return requestProcessDuration.recordCallable(() -> processBulkRequest(serviceRequestContext, aggregatedHttpRequest, bulkAPIRequestParams));
+    }
+
+    private HttpResponse processBulkRequest(final ServiceRequestContext serviceRequestContext, final AggregatedHttpRequest aggregatedHttpRequest, final BulkAPIRequestParams bulkAPIRequestParams) throws Exception {
+        requestsReceivedCounter.increment();
+        payloadSizeSummary.record(aggregatedHttpRequest.content().length());
+
+        if (serviceRequestContext.isTimedOut()) {
+            return HttpResponse.of(HttpStatus.REQUEST_TIMEOUT);
+        }
+
+        final HttpData content = aggregatedHttpRequest.content();
+        List<Map<String, Object>> bulkRequestPayloadList;
+
+        // parse the request payload
+        try {
+            bulkRequestPayloadList = jsonCodec.parse(content);
+        } catch (IOException e) {
+            LOG.error("Failed to parse the request of size {} due to: {}", content.length(), e.getMessage());
+            throw new IOException("Bad request data format.", e.getCause());
+        }
+
+        try {
+            if (buffer.isByteBuffer()) {
+                buffer.writeBytes(content.array(), null, bufferWriteTimeoutInMillis);
```

Comment:
We will probably want to support chunking. But, we can do this in a follow-on PR. Can you create a PR?

Reply:
Thanks @dlvenable. I will create a separate PR to follow up on adding the chunking functionality.